sharding/contracts/sharding_manager_test.go

```diff
@@ -1,130 +1,560 @@
+func TestNotaryRegisterInsufficientEther(t *testing.T) {
+	addr := crypto.PubkeyToAddress(mainKey.PublicKey)
+	backend := backends.NewSimulatedBackend(core.GenesisAlloc{addr: {Balance: accountBalance2000Eth}})
+	txOpts := bind.NewKeyedTransactor(mainKey)
+	txOpts.Value = notaryDepositInsufficient
+	_, _, smc, _ := deploySMCContract(backend, mainKey)
+
+	smc.RegisterNotary(txOpts)
+	backend.Commit()
+
+	notary, _ := smc.NotaryRegistry(&bind.CallOpts{}, addr)
+	numNotaries, _ := smc.NotaryPoolLength(&bind.CallOpts{})
+
+	if notary.Deposited {
+		t.Fatalf("Notary deposited with insufficient fund")
+	}
+
+	if numNotaries.Cmp(big.NewInt(0)) != 0 {
+		t.Fatalf("Incorrect count from notary pool. Want: 0, Got: %v", numNotaries)
+	}
+}
+
+func TestNotaryDoubleRegisters(t *testing.T) {
+	addr := crypto.PubkeyToAddress(mainKey.PublicKey)
+	backend := backends.NewSimulatedBackend(core.GenesisAlloc{addr: {Balance: accountBalance2000Eth}})
+	txOpts := bind.NewKeyedTransactor(mainKey)
+	txOpts.Value = notaryDeposit
+	_, _, smc, _ := deploySMCContract(backend, mainKey)
 
-	// Test deposit() function
-	// Deposit 100 Eth
-	transactOpts.Value = collatorDeposit
+	// Notary 0 registers
+	smc.RegisterNotary(txOpts)
+	backend.Commit()
+
+	notary, _ := smc.NotaryRegistry(&bind.CallOpts{}, addr)
+	numNotaries, _ := smc.NotaryPoolLength(&bind.CallOpts{})
+
+	if !notary.Deposited {
+		t.Fatalf("Notary has not registered. Got deposited flag: %v", notary.Deposited)
+	}
+
+	if numNotaries.Cmp(big.NewInt(1)) != 0 {
+		t.Fatalf("Incorrect count from notary pool. Want: 1, Got: %v", numNotaries)
+	}
+
+	// Notary 0 registers again
+	_, _ = smc.RegisterNotary(txOpts)
+	backend.Commit()
 
-	if _, err := smc.Deposit(transactOpts); err != nil {
-		t.Fatalf("Collator cannot deposit: %v", err)
+	if numNotaries.Cmp(big.NewInt(1)) != 0 {
+		t.Fatalf("Incorrect count from notary pool. Want: 1, Got: %v", numNotaries)
 	}
+
+	//ctx := context.Background()
+	//_, _ = backend.TransactionReceipt(ctx, tx.Hash())
+	////t.Log(r.CumulativeGasUsed)
+
+}
+
+func TestNotaryDeregister(t *testing.T) {
+	addr := crypto.PubkeyToAddress(mainKey.PublicKey)
+	backend := backends.NewSimulatedBackend(core.GenesisAlloc{addr: {Balance: accountBalance2000Eth}})
+	txOpts := bind.NewKeyedTransactor(mainKey)
+	txOpts.Value = notaryDeposit
+	_, _, smc, _ := deploySMCContract(backend, mainKey)
+
+	// Notary 0 registers
+	smc.RegisterNotary(txOpts)
 	backend.Commit()
 
-	// Check updated number of collators
-	numCollators, err := smc.NumCollators(&bind.CallOpts{})
+	notary, _ := smc.NotaryRegistry(&bind.CallOpts{}, addr)
+	numNotaries, _ := smc.NotaryPoolLength(&bind.CallOpts{})
+
+	if !notary.Deposited {
+		t.Fatalf("Notary has not registered. Got deposited flag: %v", notary.Deposited)
+	}
+
+	if numNotaries.Cmp(big.NewInt(1)) != 0 {
+		t.Fatalf("Incorrect count from notary pool. Want: 1, Got: %v", numNotaries)
+	}
+
+	// Fast forward 100 blocks
+	for i := 0; i < 100; i++ {
```

Comment:
Magic number alert!!! Please use the constant for lockup period and fast-forward to lockup period + 1

Reply:
It would also make it clear _why_ you are fast forwarding blocks if you use a named constant

Reply:
Ok will do.